quantum/painter/qp.h

```diff
@@ -0,0 +1,148 @@
+
+// Stream pixel data in the device's native format into the previously-set viewport
+bool qp_pixdata(painter_device_t device, const void *pixel_data, uint32_t native_pixel_count);
+
+// Set a specific pixel
+bool qp_setpixel(painter_device_t device, uint16_t x, uint16_t y, uint8_t hue, uint8_t sat, uint8_t val);
+
+// Draw a line
+bool qp_line(painter_device_t device, uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint8_t hue, uint8_t sat, uint8_t val);
+
+// Draw a rectangle
+bool qp_rect(painter_device_t device, uint16_t left, uint16_t top, uint16_t right, uint16_t bottom, uint8_t hue, uint8_t sat, uint8_t val, bool filled);
+
+// Draw a circle
+bool qp_circle(painter_device_t device, uint16_t x, uint16_t y, uint16_t radius, uint8_t hue, uint8_t sat, uint8_t val, bool filled);
+
+// Draw an ellipse
+bool qp_ellipse(painter_device_t device, uint16_t x, uint16_t y, uint16_t sizex, uint16_t sizey, uint8_t hue, uint8_t sat, uint8_t val, bool filled);
+
+// Load an image from memory. Image can be unloaded by invoking qp_close_image() below
+// - Returns NULL if unable to load
+painter_image_handle_t qp_load_image_mem(const void *buffer);
+
+// Closes an image handle when no longer in use. Returns true if successfully closed.
+bool qp_close_image(painter_image_handle_t image);
+
+// Draw an image on the device
+bool qp_drawimage(painter_device_t device, uint16_t x, uint16_t y, painter_image_handle_t image);
+bool qp_drawimage_recolor(painter_device_t device, uint16_t x, uint16_t y, painter_image_handle_t image, uint8_t hue_fg, uint8_t sat_fg, uint8_t val_fg, uint8_t hue_bg, uint8_t sat_bg, uint8_t val_bg);
+
+// Draw an animation on the device. Returns INVALID_DEFERRED_TOKEN if the animation could not be kicked off. Animation can be canceled by passing the token to qp_stop_animation() below.
+deferred_token qp_animate(painter_device_t device, uint16_t x, uint16_t y, painter_image_handle_t image);
+deferred_token qp_animate_recolor(painter_device_t device, uint16_t x, uint16_t y, painter_image_handle_t image, uint8_t hue_fg, uint8_t sat_fg, uint8_t val_fg, uint8_t hue_bg, uint8_t sat_bg, uint8_t val_bg);
+
+// Cancels an animation
+void qp_stop_animation(deferred_token anim_token);
+
+// Load a font from memory. Font can be unloaded by invoking qp_close_font() below
+// - Returns NULL if unable to load
+painter_font_handle_t qp_load_font_mem(const void *buffer);
+
+// Closes a font handle when no longer in use. Returns true if successfully closed.
+bool qp_close_font(painter_font_handle_t font);
+
+// Draw text to the display
+int16_t qp_textwidth(painter_font_handle_t font, const char *str);
+int16_t qp_drawtext(painter_device_t device, uint16_t x, uint16_t y, painter_font_handle_t font, const char *str);
+int16_t qp_drawtext_recolor(painter_device_t device, uint16_t x, uint16_t y, painter_font_handle_t font, const char *str, uint8_t hue_fg, uint8_t sat_fg, uint8_t val_fg, uint8_t hue_bg, uint8_t sat_bg, uint8_t val_bg);
```

Comment:
Personally I would prefer doxygen headers for the public API functions.

Reply:
I think the drivers API should be dynamically included in this header file as well, or it should be stated in the documentation that one have to include the drivers api specifically for the display they are using.

```c
#ifdef QUANTUM_PAINTER_ILI9341_SPI_ENABLE
#    include "qp_ili9341.h"
#endif
```

Or it will result in a warning:

```
keyboards/karlk90/dzzy/dzzy.c: In function 'keyboard_post_init_user':
keyboards/karlk90/dzzy/dzzy.c:14:15: warning: implicit declaration of function 'qp_ili9341_make_spi_device' [-Wimplicit-function-declaration]
   14 |     display = qp_ili9341_make_spi_device(240, 320, B5, B12, B7, 2, 0);
```